src/GradientPicker/index.js

```diff
@@ -0,0 +1,1459 @@
+import Element from '../Element';
+import Overlay from '../Overlay';
+import Button from '../Button';
+import SelectInput from '../SelectInput';
+import NumericInput from '../NumericInput';
+import TextInput from '../TextInput';
+import Panel from '../Panel';
+import Canvas from '../Canvas';
+import Label from '../Label';
+import { CurveSet } from '../Math/curve-set';
+import { Curve } from '../Math/curve';
+import { math } from '../Math/math';
+import { _hsv2rgb, _rgb2hsv } from '../Math/color-value';
+
+const CLASS_MULTIPLE_VALUES = 'pcui-multiple-values';
+
+const CURVE_LINEAR = 0;
+const CURVE_SPLINE = 4;
+const CURVE_STEP = 5;
+
+const REGEX_KEYS = /keys/;
+const REGEX_TYPE = /type/;
+const CLASS_GRADIENT = 'pcui-gradient';
+
+/**
+ * @name GradientPicker
+ * @augments Element
+ * @property {object} value An optional object containing the curves the picker should be initialised to
+ * @property {number} channels=3 Number of color channels; default is 3, changing to 4 adds the option to change the alpha value
+ * @classdesc Represents a gradient picker
+ */
+class GradientPicker extends Element {
+    /**
+     * Creates a new GradientPicker.
+     *
+     * @param {object} args - The arguments. Extends the Element arguments. Any settable property can also be set through the constructor.
+     */
+    constructor(args) {
+        if (!args) args = {};
+
+        super(args.dom ? args.dom : document.createElement('div'), args);
+
+        this.class.add(CLASS_GRADIENT);
+
+        this._canvas = new Canvas({ useDevicePixelRatio: true });
+        this.dom.appendChild(this._canvas.dom);
+        this._canvas.parent = this;
+        this._canvas.on('resize', this._renderGradient.bind(this));
+        this._checkerboardPattern = this.createCheckerboardPattern(this._canvas.dom.getContext('2d'));
+
+
+        // make sure canvas is the same size as the container element
+        // 20 times a second
+        this._resizeInterval = setInterval(() => {
+            this._canvas.resize(this.width, this.height);
+        }, 1000 / 20);
+
+        this._domEventKeyDown = this._onKeyDown.bind(this);
+        this._domEventFocus = this._onFocus.bind(this);
+        this._domEventBlur = this._onBlur.bind(this);
+
+        this.dom.addEventListener('keydown', this._domEventKeyDown);
+        this.dom.addEventListener('focus', this._domEventFocus);
+        this.dom.addEventListener('blur', this._domEventBlur);
+
+        this.on('click', () => {
+            if (!this.enabled || this.readOnly || this.class.contains(CLASS_MULTIPLE_VALUES)) return;
+            this._openGradientPicker();
+        });
+
+        this._channels = args.channels || 3;
+        this._value = this._getDefaultValue();
+        if (args.value) {
+            this.value = args.value;
```

Comment:
I tried using the gradient picker locally and found this line results in `setValue` being invoked which asserts because `this.STATE` is not yet defined. Think this must be moved to lower in the constructor.